### PR TITLE
feat: superchain erc20 redesign

### PR DESCRIPTION
**Pull request by agusduha**

### CODE REVIEW

File: packages/contracts-bedrock/src/L2/SuperchainERC20.sol [hunk 1]
```diff
@@ -1,49 +1,47 @@
...
-    /// @param _from   Address of the msg.sender of sendERC20 on the source chain.
-    /// @param _to     Address to relay tokens to.
-    /// @param _amount Amount of tokens to relay.
-    function relayERC20(address _from, address _to, uint256 _amount) external virtual {
-        if (msg.sender != MESSENGER) revert CallerNotL2ToL2CrossDomainMessenger();
+    /// @notice Allows the SuperchainERC20Bridge to mint tokens.
+    /// @param _to     Address to mint tokens to.
+    /// @param _amount Amount of tokens to mint.
+    function __superchainMint(address _to, uint256 _amount) external virtual onlySuperchainERC20Bridge {
+        if (_to == address(0)) revert ZeroAddress();
```
Comment (0xng) [inline packages/contracts-bedrock/src/L2/SuperchainERC20.sol]: not sure how i feel about having this check, it feels opinionated, users calling this can check it themselves if they want, tokens extending it can add it.

also minting happens async, so after the burn, so if the _to is address 0, the user will lose the tokens either way - if anything this check should be in the bridge

File: packages/contracts-bedrock/src/L2/SuperchainERC20Bridge.sol [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+    /// @param _to      Address to send tokens to.
+    /// @param _amount  Amount of tokens to send.
+    /// @param _chainId Chain ID of the destination chain.
+    function sendERC20(address _token, address _to, uint256 _amount, uint256 _chainId) external {
+        ISuperchainERC20(_token).__superchainBurn(msg.sender, _amount);
+
+        bytes memory message = abi.encodeCall(this.relayERC20, (_token, msg.sender, _to, _amount));
+        IL2ToL2CrossDomainMessenger(MESSENGER).sendMessage(_chainId, address(this), message);
+
+        emit SendERC20(_token, msg.sender, _to, _amount, _chainId);
```
Comment (0xng) [inline packages/contracts-bedrock/src/L2/SuperchainERC20Bridge.sol:39]: Let's stick to past tense like we do for all events for cohesion in the codebase, same in `relayERC20`. I know these are in present tense in the standard but we should change that

File: packages/contracts-bedrock/src/L2/SuperchainERC20.sol [hunk 3]
```diff
@@ -1,49 +1,47 @@
...
 
-        uint256 source = IL2ToL2CrossDomainMessenger(MESSENGER).crossDomainMessageSource();
+        emit SuperchainMint(_to, _amount);
+    }
 
-        _mint(_to, _amount);
+    /// @notice Allows the SuperchainERC20Bridge to burn tokens.
+    /// @param _from   Address to burn tokens from.
+    /// @param _amount Amount of tokens to burn.
+    function __superchainBurn(address _from, uint256 _amount) external virtual onlySuperchainERC20Bridge {
```
Comment (0xng) [inline packages/contracts-bedrock/src/L2/SuperchainERC20.sol:38]: the standard says these should be public, we can modify the standard for them to be external also if we prefer it

File: packages/contracts-bedrock/src/L2/SuperchainERC20.sol [hunk 4]
```diff
@@ -1,49 +1,47 @@
...
+        if (_from == address(0)) revert ZeroAddress();
```
Comment (0xng) [inline packages/contracts-bedrock/src/L2/SuperchainERC20.sol]: same with this check, not fully convinced it's needed here

File: packages/contracts-bedrock/src/L2/SuperchainERC20.sol (hunk 1 above)
Comment (0xParti) [inline packages/contracts-bedrock/src/L2/SuperchainERC20.sol]: whats the reason to enforce this revert in the standard?
Also, if you go this path, why revert here and not in the `SuperchainERC20Bridge` and save an external call?

File: packages/contracts-bedrock/src/L2/SuperchainERC20.sol (hunk 4 above)
Comment (0xParti) [inline packages/contracts-bedrock/src/L2/SuperchainERC20.sol]: same

File: packages/contracts-bedrock/src/L2/SuperchainWETH.sol [hunk 5]
```diff
@@ -5,24 +5,23 @@ pragma solidity 0.8.15;
...
 import { IL1Block } from "src/L2/interfaces/IL1Block.sol";
 import { IETHLiquidity } from "src/L2/interfaces/IETHLiquidity.sol";
+import { ISuperchainWETH } from "src/L2/interfaces/ISuperchainWETH.sol";
 
 /// @title SuperchainWETH
 /// @notice SuperchainWETH is a version of WETH that can be freely transfrered between chains
 ///         within the superchain. SuperchainWETH can be converted into native ETH on chains that
 ///         do not use a custom gas token.
-contract SuperchainWETH is WETH98, ISuperchainERC20Extensions, ISemver {
+contract SuperchainWETH is WETH98, ISuperchainWETH, ISemver {
```
Comment (0xParti) [inline packages/contracts-bedrock/src/L2/SuperchainWETH.sol:21]: I was not involved with `SuperchainWETH`, but it seems like the design has a lot of differences with the `SuperchainERC20` now. Can we push for an implementation update here?

File: packages/contracts-bedrock/test/L2/SuperchainERC20.t.sol [hunk 6]
```diff
@@ -65,159 +49,111 @@ contract SuperchainERC20Test is Test {
...
-        superchainERC20.relayERC20(_crossDomainMessageSender, _to, _amount);
+        // Call the `burn` function with the zero address
+        vm.prank(SUPERCHAIN_ERC20_BRIDGE);
+        superchainERC20.__superchainBurn({ _from: ZERO_ADDRESS, _amount: _amount });
     }
 
-    /// @notice Tests the `relayERC20` mints the proper amount and emits the `RelayERC20` event.
-    function testFuzz_relayERC20_succeeds(address _from, address _to, uint256 _amount, uint256 _source) public {
+    /// @notice Tests the `burn` burns the amount and emits the `Burn` event.
+    function testFuzz___superchainBurn_succeeds(address _from, uint256 _amount) public {
```
Comment (0xParti) [inline packages/contracts-bedrock/test/L2/SuperchainERC20.t.sol:102]: you can add the new `SuperchainBurn` event emision check here. Same for `SuperchainMint` where it belongs

**Review by 0xParti — COMMENTED** [review]
Amazing work guys <3

**Review by 0xng — COMMENTED** [review]
Looks good guys!

Run all added/modified tests with the flag `--fuzz-runs 100000` if possible so we reduce the possibilities of getting flaky errors. 

A general question I have is, we are mentioning `L2StandardBridge` in many places when the actual contract is called `L2StandardBridgeInterop`. Should we change it to that, or given that it will most likely be merged into one eventually just leave it at `L2StandardBridge`.  What do you think?

File: packages/contracts-bedrock/src/L2/SuperchainERC20.sol (hunk 3 above)
Comment (skeletor-spaceman) [reply packages/contracts-bedrock/src/L2/SuperchainERC20.sol:38]: oh good catch, I rather have them external to enforce the sender check :)

File: packages/contracts-bedrock/src/L2/interfaces/ISuperchainWETH.sol [hunk 7]
```diff
@@ -39,8 +39,6 @@ interface ISuperchainWETH {
     /// @param _dst     Address to relay tokens to.
     /// @param _wad     Amount of tokens to relay.
     function relayERC20(address _from, address _dst, uint256 _wad) external;
-
-    function __constructor__() external;
```
Comment (0xng) [inline packages/contracts-bedrock/src/L2/interfaces/ISuperchainWETH.sol:34]: what is this change?

File: packages/contracts-bedrock/scripts/ops/FeeVaultWithdrawal.s.sol [hunk 8]
```diff
@@ -65,7 +65,7 @@ contract FeeVaultWithdrawal is Script {
     }
 
     /// @notice Logs the information relevant to the user.
-    function log(uint256 _balance, address _recipient, address _vault) internal pure {
+    function log(uint256 _balance, address _recipient, address _vault) internal view {
```
Comment (0xng) [inline packages/contracts-bedrock/scripts/ops/FeeVaultWithdrawal.s.sol:68]: why this change?

File: packages/contracts-bedrock/src/L2/SuperchainERC20Bridge.sol [hunk 9]
```diff
@@ -0,0 +1,60 @@
...
+    /// @param _to      Address to send tokens to.
+    /// @param _amount  Amount of tokens to send.
+    /// @param _chainId Chain ID of the destination chain.
+    function sendERC20(address _token, address _to, uint256 _amount, uint256 _chainId) external {
+        if (_to == address(0)) revert ZeroAddress();
+
+        ISuperchainERC20(_token).__superchainBurn(msg.sender, _amount);
+
+        bytes memory message = abi.encodeCall(this.relayERC20, (_token, msg.sender, _to, _amount));
+        IL2ToL2CrossDomainMessenger(MESSENGER).sendMessage(_chainId, address(this), message);
```
Comment (0xng) [inline packages/contracts-bedrock/src/L2/SuperchainERC20Bridge.sol:37]: This function now returns the hash of the message, we can debate whether to catch the return value and include this in the event in a following PR.

**Review by 0xng — COMMENTED** [review]
Logic-wise this is approved for me. I am not approved yet because there are some changes I don't get on random files. Similarly, why was forge updated, isn't the version locked?

File: packages/contracts-bedrock/src/L2/interfaces/ISuperchainWETH.sol (hunk 7 above)
Comment (agusduha) [reply packages/contracts-bedrock/src/L2/interfaces/ISuperchainWETH.sol:34]: This is smth that was added in `develop` by the ones doing the OP contracts manager, it was added to many interfaces

This interface was not being used before so now we have to delete that function so that we dont have to implement it

File: packages/contracts-bedrock/scripts/ops/FeeVaultWithdrawal.s.sol (hunk 8 above)
Comment (agusduha) [reply packages/contracts-bedrock/scripts/ops/FeeVaultWithdrawal.s.sol:68]: Compiler error saying that the function uses `console.log` and it should be `view`

**Review by 0xng — APPROVED** [review]

**Comment by agusduha** [comment]
Closing this PR, continuing in https://github.com/defi-wonderland/optimism/pull/82
